app/src/main/java/org/mozilla/tv/firefox/tabs/TabViewHolder.kt

```diff
@@ -0,0 +1,111 @@
+import android.widget.ImageView
+import android.widget.TextView
+import android.widget.Toast
+import androidx.cardview.widget.CardView
+import androidx.recyclerview.widget.RecyclerView
+import mozilla.components.browser.session.Session
+import mozilla.components.concept.tabstray.TabsTray
+import mozilla.components.support.base.observer.Observable
+import org.mozilla.tv.firefox.R
+import org.mozilla.tv.firefox.ext.webRenderComponents
+
+/**
+ * A RecyclerView ViewHolder implementation for "tab" items.
+ */
+class TabViewHolder(
+    itemView: View,
+    private val tabsTray: WebRenderTabsTray
+) : RecyclerView.ViewHolder(itemView), Session.Observer {
+    private val cardView: CardView = (itemView as CardView).apply {
+        elevation = tabsTray.styling.itemElevation
+    }
+    private val tabView: TextView = itemView.findViewById(R.id.tabstray_url)
+    private val thumbnailView: ImageView = itemView.findViewById(R.id.tabstray_thumbnail)
+
+    private var session: Session? = null
+
+    /**
+     * Displays the data of the given session and notifies the given observable about events.
+     */
+    fun bind(session: Session, isSelected: Boolean, observable: Observable<TabsTray.Observer>) {
+        this.session = session.also { it.register(this) }
+
+        val title = if (session.title.isNotEmpty()) {
+            session.title
+        } else {
+            session.url
+        }
+
+        tabView.text = title
+
+        itemView.setOnClickListener {
+            observable.notifyObservers { onTabSelected(session) }
+        }
+
+        itemView.setOnLongClickListener {
+            val builder = AlertDialog.Builder(tabsTray.context)
+            builder.setTitle(title)
+
+            builder.setPositiveButton("Close") { dialog, _ ->
+                if (tabsTray.context.webRenderComponents.sessionManager.size > 1) {
+                    observable.notifyObservers { onTabClosed(session) }
+                } else {
+                    Toast.makeText(tabsTray.context, "You must have at least one active tab", Toast.LENGTH_SHORT).show()
+                }
+
+                dialog.dismiss()
+            }
+
+            builder.create().show()
+
+            true
+        }
+
+        if (isSelected) {
+            tabView.setTextColor(tabsTray.styling.selectedItemTextColor)
+            cardView.background = tabsTray.context.getDrawable(R.drawable.tab_tile_background_selected)
+        } else {
+            tabView.setTextColor(tabsTray.styling.itemTextColor)
+            cardView.background = tabsTray.context.getDrawable(R.drawable.tab_tile_background)
+        }
+
+        thumbnailView.setImageBitmap(session.thumbnail)
+    }
+
+    /**
+     * The attached view no longer needs to display any data.
+     */
+    fun unbind() {
+        session?.unregister(this)
+        cardView.clearFocus()
+    }
+
+    override fun onUrlChanged(session: Session, url: String) {
+        tabView.text = url
+    }
+}
+
+class TabPlusHolder(
```

Comment:
A short kdoc explaining that this is clicked to create a new session would be good for future readers.